File: Python_files/batch_generator_frame.py
import ROOT
import torch
import numpy as np

main_folder = "../"

def load_functor(num_columns):
    # Import myBatcher.C
    # ROOT.gInterpreter.ProcessLine(f'#include "{main_folder}Cpp_files/ChunkLoader.cpp"')
    ROOT.gInterpreter.ProcessLine(f'#include "{main_folder}Cpp_files/BatchGenerator.cpp"')

    # Create C++ function
    ROOT.gInterpreter.ProcessLine("""
size_t load_chunk(TMVA::Experimental::RTensor<float>& x_tensor, ROOT::RDF::RNode x_rdf,
                std::vector<std::string> cols, const size_t num_columns, 
                const size_t chunk_rows, const size_t start_row = 0, bool random_order=true) 
{
    
    // Fill the RTensor with the data from the RDataFrame
""" + f"ChunkLoader<float, std::make_index_sequence<{num_columns}>>" + """
        func(x_tensor, num_columns, chunk_rows, random_order);
    auto myCount = x_rdf.Range(start_row, start_row + chunk_rows).Count();
    x_rdf.Range(start_row, start_row + chunk_rows).Foreach(func, cols);
    return myCount.GetValue();
}
""")

class Generator:

    def __init__(self, x_rdf: ROOT.TMVA.Experimental.RTensor, columns: list[str], chunk_rows: int, 
                 batch_rows: int, num_chuncks: int=1, use_whole_file: bool=True):
        
        # Initialize parameters
        self.x_node = ROOT.RDF.AsRNode(x_rdf)
        self.columns = columns
        self.chunk_rows = chunk_rows
        self.current_chunck = 0
        self.num_chuncks = num_chuncks
        self.EoF = False

        self.num_columns = len(columns)
        self.batch_rows = batch_rows
        self.batch_size = batch_rows * self.num_columns
        self.use_whole_file = use_whole_file

        load_functor(self.num_columns)

        # Create x_tensor
        self.x_tensor = ROOT.TMVA.Experimental.RTensor("float")([self.chunk_rows, self.num_columns])    
        self.generator = ROOT.BatchLoader(self.batch_rows, self.num_columns)

    def load_chunk(self):
        start = self.current_chunck * self.chunk_rows
        print(f"{start = }")

        # Fill x_tensor and get the number of rows that were processed
        loaded_size = ROOT.load_chunk(self.x_tensor, self.x_node, self.columns, self.num_columns, 
                                     self.chunk_rows, start, False)

        #TODO: think about what to do if end of file is reached
        if (loaded_size < self.batch_rows):
            self.EoF = True

        # Create Generator
        self.generator.SetTensor(self.x_tensor, loaded_size)

    def __iter__(self):
        self.EoF = False
        self.current_chunck = 0
        self.load_chunk()

        return self

    # Return a batch when available
    def __next__(self) -> ROOT.TMVA.Experimental.RTensor:
        
        if (self.generator.HasData()):
            batch = self.generator()
            data = batch.GetData()
            data.reshape((self.batch_size,))
            return torch.Tensor(data).view(self.batch_rows, self.num_columns)

        # Load the next chunk
        self.current_chunck += 1
        if ((self.use_whole_file and not self.EoF) or (self.current_chunck < self.num_chuncks)):
            self.load_chunk()
            return self.__next__()

        raise StopIteration

File: Cpp_files/old/BatchGeneratorFrame.cpp
#include <iostream>
#include <tuple>
#include <vector>
#include <algorithm>

#include "TMVA/RTensor.hxx"
#include "ROOT/RDataFrame.hxx"
#include "ROOT/RDF/RDatasetSpec.hxx"
#include "TFile.h"

#include "ChunkLoader.cpp"

class BatchLoader
{
private:
    size_t current_row = 0, batch_size, num_rows = 0, num_columns;
    TMVA::Experimental::RTensor<float>* x_tensor;
    TMVA::Experimental::RTensor<float>* x_batch;
    bool drop_last;

    std::vector<size_t> row_order;

public:
    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // Constructors
    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    BatchLoader(const size_t batch_size, const size_t num_columns, bool drop_last=true) 
                : batch_size(batch_size), num_columns(num_columns), drop_last(drop_last) {
                    x_batch = new TMVA::Experimental::RTensor<float>({batch_size, num_columns});
                }
    
    BatchLoader(TMVA::Experimental::RTensor<float>* x_tensor, const size_t batch_size, const size_t num_columns, 
                bool drop_last=true) 
                : x_tensor(x_tensor), batch_size(batch_size), num_columns(num_columns), drop_last(drop_last) {}

    BatchLoader(TMVA::Experimental::RTensor<float>* x_tensor, const size_t batch_size, const size_t num_rows, 
                const size_t num_columns, bool drop_last=true) 
                : x_tensor(x_tensor), batch_size(batch_size), num_rows(num_rows), num_columns(num_columns), drop_last(drop_last) {}

    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // Batch function
    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    void FillBatch(std::vector<size_t> idx) {
        size_t offset;
        for (int i = 0; i < batch_size; i++) {
            offset = idx[i]*num_columns;

            // Look at std::copy

            std::copy(x_tensor->GetData() + (idx[i]*num_columns), x_tensor->GetData() + ((idx[i]+1)*num_columns), x_batch->GetData() + i*num_columns);

        }
    }

    void RandomizeOrder() {
        std::random_shuffle(row_order.begin(), row_order.end());
    }

    size_t next() 
    {
        if (current_row >= num_rows) {
            RandomizeOrder();
            current_row = 0;
        }
        return row_order[current_row++];
    }

    TMVA::Experimental::RTensor<float>* operator()()
    {
        if (current_row + batch_size <= num_rows)
        {
            std::vector<size_t> idx(batch_size);

            for (int i = 0; i < batch_size; i++) {
                idx[i] = next();
            }

            FillBatch(idx);

            return x_batch;
        }
        else
        {            
            // TODO: Implement drop_last
            return x_batch;
        }
    }
    bool HasData() {
        return (current_row + batch_size <= num_rows);}

    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // Getters and Setters
    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    void SetNumRows(size_t r) {num_rows = r;}

    void Reset(size_t num_rows) {
        this->num_rows = num_rows;
        this->current_row = 0;
    }

    void SetTensor(TMVA::Experimental::RTensor<float>* x_tensor, const size_t num_rows) {
        this->x_tensor = x_tensor;
        this->num_rows = num_rows;
        this->current_row = 0;

        row_order = std::vector<size_t>(num_rows);
        std::iota(row_order.begin(), row_order.end(), 0);
        RandomizeOrder();
    }

};


class BatchGenerator 
{
private:
    ROOT::RDataFrame* x_rdf;
    std::vector<std::string> cols;
    size_t num_columns, chunk_size, current_row=0, max_chunks, batch_size, entries;

    string file_name, tree_name;

    bool EoF = false;

    TMVA::Experimental::RTensor<float>* x_tensor;
    BatchLoader* helper;

public:
    // BatchGenerator(ROOT::RDataFrame& x_rdf, std::vector<std::string> cols, size_t chunk_size, size_t batch_size, size_t max_chunks):
    //     x_rdf(x_rdf), cols(cols), num_columns(cols.size()), chunk_size(chunk_size), max_chunks(max_chunks), batch_size(batch_size) {
        
    //     x_tensor = new TMVA::Experimental::RTensor<float>({chunk_size, num_columns});
    //     helper = new BatchLoader(batch_size, num_columns);
    // }

    BatchGenerator(string file_name, string tree_name, std::vector<std::string> cols, size_t chunk_size, size_t batch_size, size_t max_chunks):
        file_name(file_name), tree_name(tree_name), cols(cols), num_columns(cols.size()), chunk_size(chunk_size), max_chunks(max_chunks), batch_size(batch_size) {
        
        // get the number of entries in the dataframe
        TFile* f = TFile::Open(file_name.c_str());
        TTree* t = f->Get<TTree>(tree_name.c_str());
        entries = t->GetEntries();

        std::cout << entries << std::endl;

        x_rdf = new ROOT::RDataFrame(tree_name, file_name);

        x_tensor = new TMVA::Experimental::RTensor<float>({chunk_size, num_columns});
        helper = new BatchLoader(batch_size, num_columns);
    }

    void load_chunk() 
    {
        std::cout << "load_chunk " << current_row << std::endl;
        ChunkLoader<float, std::make_index_sequence<20>> func((*x_tensor), num_columns, chunk_size);

        auto myCount = x_rdf->Range(current_row, current_row + chunk_size).Count();

        x_rdf->Range(current_row, current_row + chunk_size).Foreach(func, cols);

        size_t loaded_size = myCount.GetValue();

        helper->SetTensor(x_tensor, loaded_size);

        current_row += chunk_size;
    }

    TMVA::Experimental::RTensor<float>* get_batch()
    {
        if (helper->HasData()) {
            return (*helper)();
        }

        if (current_row < entries) {
            load_chunk();
            return get_batch();
        }
        
        auto tensor = new TMVA::Experimental::RTensor<float>({0,0});
        return tensor;
    }

    bool hasData() {
        if (current_row < entries) {
            return true;
        }

        return false;
    }
};